docs: Add architecture document (#949)
deephaven/deephaven-plugins: plugins/ui/docs/architecture.md

# Architecture

> deephaven.ui is built to be a flexible and extensible [React-like](https://react.dev/learn/thinking-in-react) UI framework that can be used to create complex UIs in Python. You can create UIs using only the components provided by deephaven.ui or you can create your own components using the `@ui.component` decorator.

Reviewer: ```suggestion
deephaven.ui is a flexible and extensible [React-like](https://react.dev/learn/thinking-in-react) UI framework that can create complex UIs in Python. You can create UIs using only the components provided by deephaven.ui, or you can create your own components using the `@ui.component` decorator.
```

Reviewer: I like the assertive version

## Rendering

> When you call a function decorated by `@ui.component`, it will return an `Element` object that has a reference to the function it is decorated; that is to say, the function does _not_ run immediately. The function runs when the `Element` is rendered by the client, and the result is sent back to the client. This allows the `@ui.component` decorator to execute the function with the appropriate rendering context. The client must also set the initial state before rendering, allowing the client to persist the state and re-render in the future.

Reviewer: ```suggestion
When you call a function decorated by `@ui.component`, it will return an `Element` object that references the function it is decorated by; that is to say, the function does _not_ run immediately. The function runs when the `Element` is rendered by the client, and the result is sent back to the client. This allows the `@ui.component` decorator to execute the function with the appropriate rendering context. The client must also set the initial state before rendering, allowing the client to persist the state and re-render in the future.
```

Reviewer: Note: what I prefer as better English might not be technically accurate, but you an always dismiss that suggestion. The "by" appears to be needed.

Let's say we execute the following, where a table is filtered based on the value of a text input:

```python
from deephaven import ui

@ui.component
def text_filter_table(source, column, initial_value=""):
    value, set_value = ui.use_state(initial_value)
    ti = ui.text_field(value=value, on_change=set_value)
    tt = source.where(f"{column}=`{value}`")
    return [ti, tt]

# This will render two panels, one filtering the table by Sym, and the other by Exchange
@ui.component
def double_text_filter_table(source):
    tft1 = text_filter_table(source, "Sym", "CAT")
    tft2 = text_filter_table(source, "Exchange", "PETX")
    return ui.panel(tft1, title="Sym"), ui.panel(tft2, title="Exchange")

import deephaven.plot.express as dx

_stocks = dx.data.stocks()

tft = double_text_filter_table(_stocks)
```

> Which should result in a UI like this:

Reviewer: ```suggestion
This should result in a UI like:
```

```mermaid
sequenceDiagram
  participant U as User
  participant W as Web UI
  participant UIP as UI Plugin
  participant C as Core
  participant SP as Server Plugin

  U->>W: Run notebook
  W->>C: Execute code
  C->>SP: is_type(object)
  SP-->>C: Matching plugin
  C-->>W: VariableChanges(added=[t, tft])

  W->>UIP: Open tft
  UIP->>C: Export tft
  C-->>UIP: tft (Element)

  Note over UIP: UI knows about object tft<br/>double_text_filter_table not executed yet

  UIP->>SP: Render tft (initialState)
  SP->>SP: Run double_text_filter_table
  Note over SP: double_text_filter_table executes, running text_filter_table twice
  SP-->>UIP: Result (document=[panel(tft1), pane(tft2)], exported_objects=[tft1, tft2])
  UIP-->>W: Display Result

  U->>UIP: Change text input 1
  UIP->>SP: Change state
  SP->>SP: Run double_text_filter_table
  Note over SP: double_text_filter_table executes, text_filter_table only <br/>runs once for the one changed input<br/>only exports the new table, as client already has previous tables
  SP-->>UIP: Result (document=[panel(tft1'), panel(tft2)], state={}, exported_objects=[tft1'])
  UIP-->>W: Display Result
```

### Threads and rendering

> When a component is rendered, the render task is [submitted to the Deephaven server as a "concurrent" task](https://deephaven.io/core/pydoc/code/deephaven.server.executors.html#deephaven.server.executors.submit_task). This ensures that rendering one component does not block another component from rendering. A lock is then held on that component instance to ensure it can only be rendered by one thread at a time, a root [render context](#render-context) is set in the thread-local data, and the component is rendered.

Reviewer: wonder if numbers are helpful?
```suggestion
When a component is rendered, the render task is [submitted to the Deephaven server as a "concurrent" task](https://deephaven.io/core/pydoc/code/deephaven.server.executors.html#deephaven.server.executors.submit_task). This ensures that rendering one component does not block another component from rendering. A lock is then held on that component instance to 1) ensure it can only be rendered by one thread at a time, 2) a root [render context](#render-context) is set in the thread-local data, and  3) the component is rendered.
```

Author: Mmm the numbering here doesn't quite make sense, and maybe my sentence doesn't really either.
What I'm trying to say is "A lock is held on that component instance to ensure it can only be rendered by one thread at a time, then a root context is set in the thread-local data, and then the component is rendered". So a sequence of events, rather than all three things being the effect of the lock being held. How should I word that?

### Render context

> While rendering components, "hooks" are used to manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set, replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.

Reviewer 2: I would maybe reorder this paragraph (or add to the start of it) to start with something about what a render context is. It's a little confusing that the section on render context starts by talking about hooks

"Each component renders in its own render context which helps keep track of state and side effects". Or something like that

> While rendering components, "hooks" are used manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.

Reviewer: ```suggestion
While rendering components, "hooks" are used to manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set, replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.
```

## Communication/Callbacks

When the document is first rendered, it will pass the entire document to the client. When the client makes a callback, it needs to send a message to the server indicating which callback it wants to trigger, and with which parameters. For this, we use [JSON-RPC](https://www.jsonrpc.org/specification). When the client opens the message stream to the server, the communication looks like:

```mermaid
sequenceDiagram
  participant UIP as UI Plugin
  participant SP as Server Plugin

    Note over UIP, SP: Uses JSON-RPC
  UIP->>SP: setState(initialState)
  SP-->>UIP: documentUpdated(Document, State)

  loop Callback
    UIP->>SP: foo(params)
    SP-->>UIP: foo result

>     SP->>UIP: documentUpdated(Document, State)

Reviewer 2: This is an optional step if the callback triggered a re-render/state change. Not sure if we want to notate that in some way

      Note over UIP: Client can store State to restore the same state later
  end
```

## Communication Layers

A component that is created on the server side runs through a few steps before it is rendered on the client side:

> 1. [Element](./src/deephaven/ui/elements/Element.py) - The basis for all UI components. Generally a [FunctionElement](./src/deephaven/ui/elements/FunctionElement.py) created by a script using the [@ui.component](./src/deephaven/ui/components/make_component.py) decorator, and does not run the function until it is rendered. The result can change depending on the context that it is rendered in (e.g. what "state" is set).

Reviewer: ```suggestion
1. [Element](./src/deephaven/ui/elements/Element.py) - The basis for all UI components. Generally, a [FunctionElement](./src/deephaven/ui/elements/FunctionElement.py) created by a script using the [@ui.component](./src/deephaven/ui/components/make_component.py) decorator that does not run the function until it is rendered. The result can change depending on the context that it is rendered in (e.g., what "state" is set).
```

> 2. [ElementMessageStream](./src/deephaven/ui/object_types/ElementMessageStream.py) - The `ElementMessageStream` is responsible for rendering one instance of an element in a specific rendering context and handling the server-client communication. The element is rendered to create a [RenderedNode](./src/deephaven/ui/renderer/RenderedNode.py), which is an immutable representation of a rendered document. The `RenderedNode` is then encoded into JSON using [NodeEncoder](./src/deephaven/ui/renderer/NodeEncoder.py), which pulls out all the non-serializable objects (such as Tables) and maps them to exported objects, and all the callables to be mapped to commands that can be accepted by JSON-RPC. This is the final representation of the document that is sent to the client, and ultimately handled by the `WidgetHandler`.

Reviewer: ```suggestion
2. [ElementMessageStream](./src/deephaven/ui/object_types/ElementMessageStream.py) - The `ElementMessageStream` is responsible for rendering one instance of an element in a specific rendering context and handling the server-client communication. The element is rendered to create a [RenderedNode](./src/deephaven/ui/renderer/RenderedNode.py), which is an immutable representation of a rendered document. The `RenderedNode` is then encoded into JSON using [NodeEncoder](./src/deephaven/ui/renderer/NodeEncoder.py), which pulls out all the non-serializable objects (such as Tables) and maps them to exported objects, and all the callables to be mapped to commands that JSON-RPC can accept. This is the final representation of the document sent to the client and ultimately handled by the `WidgetHandler`.
```

> 3. [DashboardPlugin](./src/js/src/DashboardPlugin.tsx) - Client side `DashboardPlugin` that listens for when a widget of type `Element` is opened, and manage the `WidgetHandler` instances that are created for each widget.

Reviewer: ```suggestion
3. [DashboardPlugin](./src/js/src/DashboardPlugin.tsx) - Client-side `DashboardPlugin` that listens for when a widget of type `Element` is opened and manages the `WidgetHandler` instances that are created for each widget.
```

> 4. [WidgetHandler](./src/js/src/WidgetHandler.tsx) - Uses JSON-RPC communication with an `ElementMessageStream` instance to load the initial rendered document and associated exported objects. Listens for any changes and updates the document accordingly.

Reviewer 2: This sends the initial state to the server and then gets the initial render, right? Or is the state sent in a different layer

> 5. [DocumentHandler](./src/js/src/DocumentHandler.tsx) - Handles the root of a rendered document, laying out the appropriate panels or dashboard specified.

Reviewer 3: These links won't work on the external site.

They would need to be https://github.com/deephaven/deephaven-plugins....

Author: True... also there's no guarantee those files don't move later ... Should I bother linking to them? Should I link to the file on `main` branch? Or what do you think.

Reviewer 3: you could link to main, there's a risk they break. A link checker might pick that up for us though when we have that.

Author: Linking to main for now.

# Architecture

deephaven.ui is a flexible and extensible [React-like](https://react.dev/learn/thinking-in-react) UI framework that can create complex UIs in Python. You can create UIs using only the components provided by deephaven.ui, or you can create your own components using the `@ui.component` decorator.

## Components

Components are re-usable pieces of UI that can be composed together to create complex UIs. Each component defines it's own logic and appearance. Components can be simple, like a button, or complex, like a table with controls for filtering and sorting. Components can also be composed of other components, allowing for complex UIs to be built up from simpler pieces.

Components are created using the `@ui.component` decorator. This decorator takes a function that returns a list of components, and returns a new function that can be called to render the component. The function returned by the decorator is called a "component function".

```python
from deephaven import ui

@ui.component
def my_button():
    return ui.button("Click me!")
```

Once you have declared a component, you can nest it into another component.

```python
@ui.component
def my_app():

>     return ui.flex(ui.text("Hello, world!"), my_button(), direction="column")

Reviewer 3: to be runnable, would also need to call and assign my_app()

Let's say we execute the following, where a table is filtered based on the value of a text input:

```python
from deephaven import ui

@ui.component
def text_filter_table(source, column, initial_value=""):
    value, set_value = ui.use_state(initial_value)
    ti = ui.text_field(value=value, on_change=set_value)
    tt = source.where(f"{column}=`{value}`")
    return [ti, tt]

# This will render two panels, one filtering the table by Sym, and the other by Exchange
@ui.component
def double_text_filter_table(source):
    tft1 = text_filter_table(source, "Sym", "CAT")
    tft2 = text_filter_table(source, "Exchange", "PETX")
    return ui.panel(tft1, title="Sym"), ui.panel(tft2, title="Exchange")

> import deephaven.plot.express as dx

Reviewer 3: I would put imports at the top like a sane person.

  SP-->>UIP: Result (document=[panel(tft1), pane(tft2)], exported_objects=[tft1, tft2])
  UIP-->>W: Display Result
  deactivate UIP

  U->>UIP: Change text input 1
  activate UIP
  UIP->>SP: Change state
  SP->>SP: Run double_text_filter_table
  Note over SP: double_text_filter_table executes, text_filter_table only <br/>runs once for the one changed input<br/>only exports the new table, as client already has previous tables
  SP-->>UIP: Result (document=[panel(tft1'), panel(tft2)], <br/>state={}, exported_objects=[tft1'])
  UIP-->>W: Display Result
  deactivate UIP
```

### Threads and rendering

When a component is rendered, the render task is [submitted to the Deephaven server as a "concurrent" task](https://deephaven.io/core/pydoc/code/deephaven.server.executors.html#deephaven.server.executors.submit_task). This ensures that rendering one component does not block another component from rendering. A lock is then held on that component instance to ensure it can only be rendered by one thread at a time. After the lock is acquired, a root [render context](#render-context) is set in the thread-local data, and the component is rendered.

### Render context

> Each component renders in its own render context, which helps keep track of state and side effects. While rendering components, "hooks" are used to manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set, replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.

Reviewer: should you define hooks more explicitly before discussing their magic?

Author: Hmm they likely deserve their own instructional page, in the `hooks` folder. Right now I have pages for specific hooks (`use_state`, `use_effect`, `use_memo`) but having a top level hooks pages similar to the [React docs](https://react.dev/reference/react/hooks) would be handy.
I'll capture this as part of https://github.com/deephaven/deephaven-plugins/issues/996

Reviewer: sounds good to me!

  SP-->>UIP: Result (document=[panel(tft1), pane(tft2)], exported_objects=[tft1, tft2])
  UIP-->>W: Display Result
  deactivate UIP

  U->>UIP: Change text input 1
  activate UIP
  UIP->>SP: Change state
  SP->>SP: Run double_text_filter_table
  Note over SP: double_text_filter_table executes, text_filter_table only <br/>runs once for the one changed input<br/>only exports the new table, as client already has previous tables
  SP-->>UIP: Result (document=[panel(tft1'), panel(tft2)], <br/>state={}, exported_objects=[tft1'])
  UIP-->>W: Display Result
  deactivate UIP
```

### Threads and rendering

When a component is rendered, the render task is [submitted to the Deephaven server as a "concurrent" task](https://deephaven.io/core/pydoc/code/deephaven.server.executors.html#deephaven.server.executors.submit_task). This ensures that rendering one component does not block another component from rendering. A lock is then held on that component instance to ensure it can only be rendered by one thread at a time. After the lock is acquired, a root [render context](#render-context) is set in the thread-local data, and the component is rendered.

### Render context

> Each component render in its own render context, which helps keep track of state and side effects. While rendering components, "hooks" are used to manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set, replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.

Reviewer 2: ```suggestion
Each component renders in its own render context, which helps keep track of state and side effects. While rendering components, "hooks" are used to manage state and other side effects. The magic part of hooks is they work based on the order they are called within a component. When a component is rendered, a new context is set, replacing the existing context. When the component is done rendering, the context is reset to the previous context. This allows for nested components to have their own state and side effects, and for the parent component to manage the state of the child components, re-using the same context when re-rendering a child component.
```
